src/dlr_relayvm.cc

```diff
@@ -0,0 +1,290 @@
+}
+
+void RelayVMModel::UpdateInputs() {
+  const int kNumArgs = GetNumInputs() + 1;
+  TVMValue values[kNumArgs];
+  int type_codes[kNumArgs];
+  auto arg_setter = tvm::runtime::TVMArgsSetter(values, type_codes);
+  arg_setter(0, ENTRY_FUNCTION);
+  for (int i = 0; i < inputs_.size(); i++) {
+    arg_setter(i + 1, inputs_[i]);
+  }
+
+  tvm::runtime::PackedFunc set_input = vm_module_->GetFunction("set_input");
+  tvm::runtime::TVMRetValue rv;
+  set_input.CallPacked(tvm::runtime::TVMArgs(values, type_codes, kNumArgs),
+                       &rv);
+}
+
+void RelayVMModel::Run() {
+  // Invoke inference
+  UpdateInputs();
+  tvm::runtime::PackedFunc invoke = vm_module_->GetFunction("invoke");
+  output_ref_ = invoke(ENTRY_FUNCTION);
+  UpdateOutputs();
+}
+
+void RelayVMModel::UpdateOutputs() {
+  outputs_.resize(num_outputs_);
+  if (output_ref_->IsInstance<tvm::runtime::ADTObj>()) {
+    auto adt = tvm::runtime::Downcast<tvm::runtime::ADT>(output_ref_);
+    for (int i = 0; i < adt.size(); i++) {
+      outputs_[i] = tvm::runtime::Downcast<tvm::runtime::NDArray>(adt[i]);
+    }
+  } else if (output_ref_->IsInstance<tvm::runtime::NDArray::ContainerType>()) {
+    outputs_[0] = tvm::runtime::Downcast<tvm::runtime::NDArray>(output_ref_);
+  } else {
+    throw dmlc::Error("Invalid output_ref format!");
+  }
+}
+
+void RelayVMModel::GetOutput(int index, void *output) {
+  CHECK_LT(index, num_outputs_) << "Output index is out of range.";
+  auto out_array = outputs_[index];
+  DLTensor output_tensor;
+  output_tensor.data = output;
+  output_tensor.ctx = ctx_;
+  output_tensor.ndim = out_array->ndim;
+  output_tensor.dtype = out_array->dtype;
+  output_tensor.shape = out_array->shape;
+  output_tensor.strides = nullptr;
+  output_tensor.byte_offset = 0;
+  out_array.CopyTo(&output_tensor);
+}
+
+void RelayVMModel::GetOutputShape(int index, int64_t *shape) const {
+  CHECK_LT(index, num_outputs_) << "Output index is out of range.";
+  std::memcpy(shape, outputs_[index]->shape,
+              sizeof(int64_t) * outputs_[index]->ndim);
+}
+
+void RelayVMModel::GetOutputSizeDim(int index, int64_t *size, int *dim) {
+  CHECK_LT(index, num_outputs_) << "Output index is out of range.";
+  *size = 1;
+  auto arr = outputs_[index];
+  *size = std::accumulate(arr->shape, arr->shape + arr->ndim, 1,
+                          std::multiplies<int64_t>());
+  *dim = arr->ndim;
+}
+
+const char *RelayVMModel::GetOutputType(int index) const {
+  CHECK_LT(index, num_outputs_) << "Output index is out of range.";
+  return output_types_[index].c_str();
+}
+
+const char *RelayVMModel::GetBackend() const { return "relayvm"; }
+
+void RelayVMModel::SetNumThreads(int threads) {
+  throw dmlc::Error("Not Implemented!");
+}
+
+void RelayVMModel::UseCPUAffinity(bool use) {
+  throw dmlc::Error("Not Implemented!");
+}
+
+bool RelayVMModel::HasMetadata() const { return !metadata_.is_null(); }
+
+const char *RelayVMModel::GetOutputName(const int index) const {
```

Comment:
int index

Reply:
This is an overridden function. Will fix this as part of refactor.